modify timing output to add output_time and total_wall_time

rjleveque opened the pull request:
@mandli: This adds two new columns to the csv file, for `output_time` and `total_wall_time`.

Note this required adding `tick_clock_start` to amr_module.f90

Results can be plotted with https://github.com/rjleveque/visclaw/blob/plot_timing_stats/src/python/visclaw/plot_timing_stats.py
which still needs some work.

rjleveque committed:
modify timing output to add output_time and total_wall_time

required adding new module variable to amr_module.f90

mandli commented on src/2d/valout.f90:
```diff
@@ -382,15 +384,24 @@ subroutine valout(level_begin, level_end, time, num_eqn, num_aux)
 
-    call cpu_time(t_CPU_overall)
-    write(out_unit, timing_line) t_CPU_overall, &
+    if (time == t0) then
```
Can this be set in the `amr_module` or something similar?  I am worried that the comparison might not work or that this will get mucked up on a restart.

mandli commented:
Just a small comment on the setting of the new variables in `valout`.  Can you also make the same changes to the 1D and 3D?

rjleveque committed:
add examples/advection_2d_inflow/plot_timing_stats.py to show how to plot timings

rjleveque commented:
1. At the moment updating things on restarting isn't working well and needs some work.  The reason for my `if (time == t0)` in  `valout` is that it is using timers set in `tick`, which isn't called until after `valout` is called at the initial time.  I think `t0` is generally unchanged on a restart, so the currently logic should be ok, but the `csv` file isn't updated properly after a restart and plots look wonky.  All of this needs some more thought.

2. I'll take a look at 1D and 3D and make it consistent with 2D.

3. What's the plan for updating `valout` in GeoClaw?  Carry these changes over there, or remove `valout` from GeoClaw and rewrite things to get topography from the `fort.a` files?

4. For now I put an updated `plot_timing_stats.py` in one example directory, `examples/advection_2d_inflow`.  Eventually this should be generalized and go into visclaw, but that's lower priority than getting `valout` changes finalized and merged in.

mandli commented:
1. Hmm, that makes sense but couldn't we still initialize the timing in `amr_module` or `amr2`?
1. Thanks!
1. I have a version of this in GeoClaw as well, just need to carry over the timing stuff.
1. 👍

rjleveque commented:
It's not so easy to do it the same way in 1d and 3d, since the timing stuff is organized differently there and not set in `tick` as in 2d.  

I am also noticing that what is checkpointed for restarts is now very different in 2d than in 1d and 3d,  both the timing stuff and also the new maxgr etc. that @mjberger added in clawpack/amrclaw#215 in 2d but not in 1d or 3d.  

This is going to need a bigger cleanup to sync things back up again.   Should we tackle this now?

rjleveque committed:
modify src/1d/valout.f90 and tick, amr_module to print more timing info

Note that timing is done differently in 1d than 2d, does not yet support restart

rjleveque commented:
I modified the 1d code so at least it prints the same things and has the modification to `amr_module`, but not yet fully implemented with timing started in `tick` to work with restarts.  I can do the same in 3d if this seems ok.

In testing this I noticed the `clock_rate = 1000` so the call to `system_call` is giving back an integer number of milliseconds. For 1d runs in the examples directory the full run time is so small that some of the plots come out looking nonsensical, but if you increase `tfinal` substantially, or use a finer grid, they look ok.  The same `examples/advection_2d_inflow/plot_timing_stats.py` can now be run in a 1d example directory.

mandli commented:
Should we change the clock_rate to `1000` then?

rjleveque commented:
I think `clock_rate` is set by the call to `system_clock`.

mandli commented:
I say we just go with this when the 3D stuff is there and I will raise an issue.

rjleveque committed:
modify src/3d/valout.f90 and tick, amr_module to print more timing info

As in 1d, this needs further work to allow restart

rjleveque commented:
@mandli: Is this ready to merge?

mandli commented:
Looks good to me.
